Torch instance clip

binh234 opened the pull request:
close #13639 
close #13640

hmahmood24 commented on ivy/functional/frontends/torch/tensor.py:
```diff
@@ -683,6 +683,15 @@ def clamp_(self, min=None, max=None, *, out=None):
+    def clip_(self, min=None, max=None):
+        self._ivy_array = self.clip(min=min, max=max).ivy_array
+        return self
+
```
Since `clip` is also an alias for `clamp`, and `clip_` is an alias for `clamp_`, can you kindly add them [here](https://github.com/unifyai/ivy/blob/a86a2fb82dfcf45870a178d723b0df2eca728b75/ivy/functional/frontends/torch/tensor.py#L891-L893) as well in this file?

hmahmood24 reviewed (changes requested):
Hey @binh234, thanks so much for an awesome contribution! I have requested a minor modification for you to address. Also, the tests seem to be failing, can you kindly look into them [here](https://github.com/unifyai/ivy/actions/runs/4603801293/jobs/8134134626?pr=13652#step:3:133) and get them fixed?

Regarding your suggestions, even though they seem plausible to me, I'm not so sure if we should be changing multiple methods like that. We can instead perform a `git blame` and ask the original author their reason behind implementing that logic rather than changing it without fully aware of its consequences as such. 

Kindly ask for a re-review from me once you've addressed the changes to the PR and fixed the tests. Thank you! 🙂

binh234 commented:
@hmahmood24, as you suggested for alias methods, I have removed my implementation and used the same helper function for testing to avoid redundancy.
Currently, the clamp method will fail when tested on complex tensors, as torch doesn't support clamp operation on these types of tensors. I also added `complex` to the `unsupported_dtypes` wrapper to make the tests pass.
Just one more thing to consider, about the `min > max` logic, I still think we should cover this in the frontend method.

SaabriinMo commented on ivy/functional/frontends/torch/tensor.py:
```diff
@@ -694,13 +694,13 @@ def tril(self, diagonal=0):
-    @with_unsupported_dtypes({"1.11.0 and below": ("bfloat16", "float16")}, "torch")
+    @with_unsupported_dtypes({"1.11.0 and below": ("float16", "complex")}, "torch")
     def clamp(self, min=None, max=None):
         if min is not None and max is not None and ivy.all(min > max):
```
basically, there's two reasons for this:
1. In the documentation (under notes), it says that if the min is greater than the max, then it would set all the values to max. So i had to add this in the function.
2. if i didnt add this case, it would throw an error as some of the test cases have examples where min > max.

Here's the documentation Im referring to (its under notes): 
https://pytorch.org/docs/stable/generated/torch.clamp.html

binh234 replied:
Thanks for the explanation! Here are my opinions:

1. The frontend method and instance method should have the same behavior. If you only add the logic to the instance method, the frontend method will behave differently when `min > max`,  this is not what we want! Also, the note you gave me is about the torch.clamp method, hence the logic for `min > max` should be handled in the frontend method, not the instance method.
If you still need to cover min > max cases, we should perform a `git blame` and ask the author of the clamp frontend method for changes instead, like what we are doing now
2. I tried to remove the `min > max` part of your code, and all the tests still passed. When I deep dive to reproduce your errors, I found that the errors were related to the hypothesis library, when it tried to sample integer arrays, then threw something like this:
```bash
hypothesis.errors.InvalidArgument: Cannot have max_value=-1 < min_value=1
```
This error will appear when you try to sample min array from range [-25, 0] and max array from range [1, 25]. I noticed that you have changed to sample min array from range [0, 25] and max array from range [26, 50], then the error disappears, like magic :)
3. Finally, ivy.all is not the right operator for this case, e.g. when you pass min = tensor([0.5, -0.5]) and max = tensor([-0.5, 0.5], ivy.all will return False and then call the clamp frontend method

SaabriinMo replied:
1. the first statement is true, i was going to get round to that but didn't have the time :'(, thanks for pointing that out. i was wondering why it wasn't implemented in the frontend method so i added it in the instance method then come back to it at a later date and bring this up! 
2. I see, i thought ivy.all was best when completing the clamp function which was why i implemented it! 
3. i dont remember changing the min/max array, i thought i kept it as min=[-25,0] and max=[1,25]. i would need to change that!

SaabriinMo replied:
i see that you're removing bfloat16! Please don't remove that as it has caused hypothesis errors regarding this! thank you.

binh234 replied:
Can you give me the error details? I have tested it several times but I can't see any errors, maybe the hypothesis team has fixed that error!
As you can see in Ivy docs [here](https://lets-unify.ai/docs/ivy/overview/deep_dive/data_types.html#supported-and-unsupported-data-types), the decorator `@with_unsupported_dtypes` and `@with_supported_dtypes` are used to flag the data types which this particular function does and does not support respectively for the associated backend. If torch supports bfloat16 for clamp, we should not add bfloat16 as an unsupported dtype.

SaabriinMo replied:
E       RuntimeError: "clamp_scalar_cpu" not implemented for 'Half'

E       TypeError: expected TensorOptions(dtype=float, device=cpu, layout=Strided, requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt)) (got TensorOptions(dtype=c10::BFloat16, device=cpu, layout=Strided, requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt)))
E       Falsifying example: test_torch_instance_clamp(
E           dtype_and_x=(['bfloat16'], [array([[-1]], dtype=bfloat16)]),
E           as_variable=False,
E           native_array=False,
E           class_=ivy.functional.frontends.torch.tensor.Tensor,
E           method_name='clamp',
E       )
E       

this was the error i came across when i was working with clamp

binh234 replied:
If you got `RuntimeError: "clamp_scalar_cpu" not implemented for 'Half'`, it seems that you're using float16/half on CPU which does not support many of the operators ([here](https://github.com/pytorch/pytorch/issues/67684)). `Half` represents float16, and `BFloat16` represents bfloat16.

SaabriinMo replied:
yh that makes sense since i was using cpu rather than gpu! I understood that it was regarding float16 but i was unsure if it was bfloat16 or just float16! Thanks :)

hmahmood24 commented:
Hey @binh234, thanks so much for the proactive work! I agree we should not handle any edge cases in the instance method itself since that will definitely not be the best approach to unifying the behavior of identical functions/methods. Any extra logic should reside in a common interface between all methods and functions, hence the frontend function seems like the right place to move the `min > max` logic to. Please feel free to make this change in this PR so that we can go ahead and merge this into `master` 🙂 Thanks again!

binh234 commented:
Hey @hmahmood24, actually, the current implementation looks good to me.

If you take a look at [Tensorflow](https://www.tensorflow.org/api_docs/python/tf/clip_by_value)/[JAX](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.clip.html)/[Numpy](https://numpy.org/doc/stable/reference/generated/numpy.clip.html) docs, we can see that they don't perform any checks to ensure that `min < max`. On the other hand, clip can be considered as a composition of minimum and maximum, from [torch docs](https://pytorch.org/docs/stable/generated/torch.clamp.html). If we want to get similar outputs to torch, we can perform `ivy.minimum(ivy.maximum(x, min), max)`. But, this will raise some performance considerations, as Numpy said: `np.clip is equivalent to but faster than np.minimum(a_max, np.maximum(a, a_min))`. So I decide to keep the current implementation, as it is already optimized.

I have done some refactoring instead, including:
- Added `complex` to unsupported dtypes
- Removed ivy.array calls, as the `@to_ivy_arrays_and_back` decorator already do it for us.
- Changed clip to an alias method of clamp
- Tested torch.clamp on both `int` and `float` inputs.

hmahmood24 commented:
Hey @binh234, great work! 😃 Everything looks good to merge now. Thanks so much for an amazing contribution! 🚀

bhack commented:
In the case you are interested there is an upstream ticket now https://github.com/pytorch/pytorch/issues/99563
